### ISSUE TITLE
Add a mechanism for untrusted users to offload builds of fixed-output derivations

### ISSUE THREAD

**Issue by catern** [NixOS/nix#1734]
My problem is simple: I am operating in environments where sandboxed fixed-output derivations to fetch sources cannot work, even if the sandbox is nominally allowed to access the network. Such situations include:
- Building behind a corporate firewall that requires user-specific credentials to bypass
- Building on machines without internet access
- Building against sources that require some user-specific credentials to download

Because these are fixed-output derivations, there is no need (in terms of correctness) for these builds to be sandboxed. So it seems to me that there is a straightforward and general way to solve the problem in all of those cases: We add a mechanism whereby a user building a derivation can handle building the fixed-output derivations themselves, outside the sandbox.

There are a number of ways that could be implemented, some better than others:
1. nix-store --realise --unsandboxed-fixed-output-derivations (buildPaths in store-api) could be taught to build fixed-output derivations "manually" outside the store API, directly forking the builder off from the current process and adding the results to the store with addToStore.
2. Add a new setting "untrusted-binary-caches" which can be specified by untrusted users to list some additional binary caches which should be used only for fixed-output derivations.
(This would not truly solve the problem, because a user of this mechanism would need to pre-process derivations before they are built to discover the fixed-output derivations they need to handle, and if they were already pre-processing derivations then they might as well immediately build the fixed-output derivations outside the sandbox and call addToStore.)
3. Add a new feature "untrusted-build-hook" which can be specified by untrusted users to name a script which should be executed as a build hook for fixed-output derivations.
4. Add a new feature to the store-api where a callback can be passed to buildPaths, which is called for each fixed-output derivation, and can either accept the derivation or decline it, similar to a build hook.
(In future work, this callback could be extended to be called for all derivations (for trusted users) and replace the build hook mechanism entirely.)

These are ranked in order of increasing difficulty to implement. Of these options, I think 4 is best in the long run, but I would be content with any of them except for 2. Any other mechanism that allows automatically dispatching the build of fixed-output derivations to the user invoking the build would also be acceptable.

I personally am extremely interested in implementing any kind of fix for this issue, as it is a major blocker for ergonomic usage of Nix for my users. Currently I just manually work around it with various pre-processing scripts involving "nix-prefetch-*" or "nix-store --add". As such, as soon as I get some guidance on what the appropriate solution is, I'll immediately start implementing it and deploy it on my Nix cluster.

**Comment by shlevy** [comment]
So, I'm generally not a fan of the fact that fixed output drvs hash differently from things just added directly to the store, but one thing you can do without any changes to nix is just write a new fixed output derivation yourself with the same name and output hash, that just copies `${./the-actual-fixed-output-thing}` to `$out`

**Comment by catern** [comment]
> I'm generally not a fan of the fact that fixed output drvs hash differently from things just added directly to the store

Not sure what you mean. Like I said, my current workaround is just using "nix-prefetch-url" and "nix-store --add", and that works fine. Fixed-output derivations *are* stored under the same path as things just added to the store, AFAICT?

**Comment by catern** [comment]
My current workaround is this script, which walks over a derivation to find the fixed-output derivations it needs, and then runs the builder and adds the output to the store: https://github.com/catern/nix-utils/blob/master/fixedout.py

**Comment by FRidh** [comment]
Having one machine that has internet access and serves a binary cache is not an option for you?

**Comment by catern** [comment]
Yes, I have proxies set up which allow unauthenticated access to cache.nixos.org, but that's not a complete solution; a lot of packages in Nixpkgs are not built by Hydra, so the build must take place locally, and without internet access that can't be done.

Also, even if everything in Nixpkgs was built by Hydra, I'd still want to be able to (say) upgrade a package locally and rebuild it locally without having to push it out to Nixpkgs or other central services.

**Comment by catern** [comment]
I still care about this.

**Comment by catern** [comment]
I still care about this (and my users are still using that Python script workaround I linked above)

**Comment by yueyinqiu** [comment]
I'm wondering if there's a better approach to this nowadays? Deploying a public NixOS server in a network-restricted environment still seems like a tough problem. A native solution can be very useful.

https://discourse.nixos.org/t/best-practices-for-air-gapped-nixos-servers-with-untrusted-users/79498

And the solution I perfer might be something like a `nix copy-build` command—one that allows me to copy a package to a remote host while remote can build some stuffs when necessary. Ideally, as much as possible happens locally. However, in untrusted scenarios (e.g., local building is required, but the user isn't in trusted-users), the build step can be offloaded to the remote machine. Crucially, FODs (Fixed-Output Derivations) and binary cache downloads should always happen locally, as they are trusted.
